RecordArg, inverse of RecordArgs trait

beezee opened the pull request:
This adds a RecordArg trait, which is essentially the inverse of the RecordArgs trait. Specifically, instead of allowing forward from "method" to "methodRecord" where method can be passed named parameters that will be rolled into a record for the forwarded call, this allows forwarding from "methodRecord" to "method", where the signature of method is used to access required parameters from a record passed to the original call.

This allows a more typesafe application of functions that may take a series of similarly typed parameters, so that parameter order cannot be mistaken in such a way as to produce errors or incorrect behavior at runtime.

EG - https://gist.github.com/beezee/fe604473c557564f9a6c1be06bc0960d

Re-open of #602 after correcting commit history and cleaning up a stray import and definition.

fthomas commented on core/src/main/scala/shapeless/records.scala:
```diff
@@ -152,4 +193,17 @@ class RecordMacros(val c: whitebox.Context) {
+
+  def mkParamsImpl[L <: HList](method: MethodSymbol, rec: Expr[L]): List[Tree] = {
+    def mkElem(keyTpe: Type, value: Tree): Tree =
+      q"implicitly[shapeless.ops.hlist.Selector[${rec.actualType}, ${mkFieldTpe(keyTpe, value.tpe)}]].apply($rec)"
```
@beezee A general remark when writing macros: Non-interpolated terms and types in quasiquotes (like `implicitly` and `Selector`) should be fully qualified to avoid macro hygiene issues. So in this case you should use `_root_.scala.Predef.implicitly` and `_root_.shapeless.ops.hlist.Selector`.

beezee replied:
Thanks @fthomas that makes sense. I just pushed an update accordingly.

milessabin commented:
Awesome! Thanks!

I think we need a different name for this though ... `RecordArg` is a bit too close to `RecordArgs` for comfort. I guess that `RecordArgs` should probably have been named `ToRecordArgs` in which case it would make sense to call this `FromRecordArgs`?

Also, I think it would make sense to add a positional variant of this (ie. corresponding to `ProductArgs` rather than `RecordArgs`) ... given that that would most likely share some implementation with this (or if not, at least have a very similar flavour) would you mind adding that too?

beezee commented:
@milessabin makes sense. I renamed mine as suggested, leaving the original as it was. It occurred to me there's no reason to limit to a single param list, so I adjusted to support it.

I also added FromProductArgs. There's one thing that's bugging me about it right now, which is that it currently ignores extraneous elements on the incoming HList. That is, if the first n elements align with a target method signature, the application is successful and any additional elements are just ignored.

I think it would be preferable to make this a compile time error, but I struggled to get a stable value representing the runtime length of the incoming HList expression for comparison to the target method. Do you know of a way to do this?

beezee commented:
Weird I did not see this test failure locally. Seems to be a serialization related issue. Looking into it.

milessabin commented:
LGTM! Many thanks ... merging :-)
